### PR TITLE
Add OWL-Datalog reasoner: Horn fragment approximation of SROIQ(D)

### PR DESCRIPTION
**Pull request by bplatz**
## Summary

Implements OWL-Datalog reasoner, extending OWL 2 RL with additional constructs while maintaining Datalog compatibility. This provides a rule-materialization approach approximating parts of SROIQ(D) - more expressive than OWL 2 RL but incomplete for full OWL DL.

### Extensions Beyond OWL 2 RL

- **Complex intersections with restrictions**: Full support for equivalentClass definitions containing multiple restrictions in intersections
- **Unions compiled as rule alternatives**: Union classes generate separate Datalog rules for each disjunct
- **Property chains with inverses**: Supports chains including inverse properties in any position with double-inverse normalization
- **Forward entailment for hasValue/allValuesFrom**: Derives property values from class membership and types from universal restrictions
- **Multiple restrictions on same property**: Handles complex patterns like `Item ∩ ∃p.A ∩ ∃p.B ∩ ∃p.{v}`

### Technical Implementation

- Queries ontology with depth-6 traversal to capture nested structures
- Compiles OWL constructs to Datalog rules for forward-chaining inference
- Handles inline property chains in restrictions (e.g., `∃(R₁ ∘ R₂⁻).C`)
- Uses `util/sequential` to handle JSON-LD list collapsing for single-item chains

## Current Limitations

1. **Data property hasValue with typed literals**: Backward inference from typed literal values to class membership not yet implemented
2. **Property chains with separate axioms**: allValuesFrom restrictions don't work when property has chain axiom defined separately (works for inline chains)
3. **Qualified cardinalities**: Parsed but not reasoned over

## Testing

- 19 tests, 50 assertions, 0 failures
- Comprehensive test suites for core functionality, edge cases, and restriction features
- Known limitations documented with test cases

## Files Changed

- `src/fluree/db/reasoner/owl_datalog.cljc`: Core reasoner implementation
- `src/fluree/db/flake/reasoner.cljc`: Integration with reasoning framework
- `test/fluree/db/reasoner/owl_datalog_test.clj`: Core functionality tests
- `test/fluree/db/reasoner/owl_datalog_edge_cases_test.clj`: Complex scenario tests
- `test/fluree/db/reasoner/owl_datalog_restrictions_test.clj`: Restriction feature tests
- `docs/owl-datalog.md`: Comprehensive documentation

## Usage

```clojure
(def reasoned-db @(fluree/reason db :owl-datalog))
```

The reasoner generates materialized inferences stored alongside original data, providing fast query performance over derived triples.

## Documentation

See `docs/owl-datalog.md` for detailed documentation including examples, architecture details, and comparison with other reasoners.

### CODE REVIEW

File: docs/owl-datalog.md [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+
+```turtle
+# Forward entailment works:
+KilogramMeasurement(?x) → hasUnit(?x, kg)
+
+# Backward inference doesn't work:
+hasQualityScore(?x, "95"^^xsd:int) ↛ HighQuality(?x)
+```
+
+**Workaround**: Use untyped literals or object properties where possible.
```
Comment (dpetran) [inline docs/owl-datalog.md]: What is an "untyped literal" in this context?

File: src/fluree/db/util.cljc [hunk 2]
```diff
@@ -254,9 +254,10 @@
 (defn of-type?
   "Returns true if the provided json-ld node is of the provided type."
   [jsonld rdf-type]
-  (->> jsonld
-       get-types
-       (some #(= % rdf-type))))
+  (let [types (get-types jsonld)]
+    (if (sequential? types)
```
Comment (dpetran) [inline src/fluree/db/util.cljc]: I think it would be simpler just to use `sequential` here:

```
(->> jsonld
       get-types
       sequential
       (some #(= % rdf-type)))
```

File: src/fluree/db/reasoner/owl_datalog.cljc [hunk 3]
```diff
@@ -607,11 +1051,14 @@
 (defmethod to-datalog ::cax-eqc
   [_ inserts owl-statement all-rules]
   (let [c1 (util/get-id owl-statement) ;; the class which is the subject
+        equiv-class-val (get owl-statement const/iri-owl:equivalentClass)
+        unwrapped (util/unwrap-list equiv-class-val)
+        ;; Ensure unwrapped is always a sequence for group-by
+        unwrapped-seq (if (sequential? unwrapped) unwrapped [unwrapped])
```
Comment (dpetran) [inline src/fluree/db/reasoner/owl_datalog.cljc]: We have `util/sequential` for this, though I think it would be better for `unwrap-list` can be refactored to always return a sequence, since other callsites assume that it does.

File: src/fluree/db/reasoner/owl_datalog.cljc [hunk 4]
```diff
@@ -607,11 +1051,14 @@
...
         ;; combine with all other equivalent classes for a set of 2+ total classes
         {:keys [classes intersection-of union-of one-of
                 has-value some-values all-values
-                max-cardinality max-qual-cardinality]} (->> (get owl-statement const/iri-owl:equivalentClass)
-                                                            util/unwrap-list
+                max-cardinality max-qual-cardinality]} (->> unwrapped-seq
```
Comment (dpetran) [inline src/fluree/db/reasoner/owl_datalog.cljc]: No need for the thread when we only have a single expression.

File: src/fluree/db/reasoner/owl_datalog.cljc [hunk 5]
```diff
@@ -552,7 +976,28 @@
...
+                                                       acc**))
+                                                   acc*
+                                                   nested-unions))
+
+                                         :else acc*))
+                                     []
+                                     unions)
+           {:keys [has-value]} (group-by equiv-class-type unions)
            restrictions      (cond->> []
                                has-value (into (has-value-condition "?y" has-value)))
```
Comment (dpetran) [inline src/fluree/db/reasoner/owl_datalog.cljc]: I'm not sure why we need to call `into` here with an empty vector, we'll never add any data to the result of `has-value-condition` regardless of the value of `has-value`.

File: src/fluree/db/reasoner/owl_datalog.cljc [hunk 6]
```diff
@@ -526,6 +888,66 @@
...
+                                                                   property "?y"}]
+                                                        "insert" {"@id"   "?y"
+                                                                  "@type" target-class}})]
+                                            (conj acc* [(str rule-class "(owl:allValuesFrom-forward-" property ")") rule]))
+                                          acc*)))
+                                    []
+                                    all-values)
+
+           ;; Generate forward entailment rules for hasValue in intersections
+           has-value-rules (reduce (fn [acc* has-val-restriction]
```
Comment (dpetran) [inline src/fluree/db/reasoner/owl_datalog.cljc]: I think this function would be more readable if we extract the `has-value-rules` and `all-values-rules` into their own functions.

File: test/fluree/db/reasoner/owl_datalog_test.clj [hunk 7]
```diff
@@ -0,0 +1,401 @@
...
+                                                      "@type" "ex:Formulation"}}}
+
+          db-with-data @(fluree/update db-with-ontology instance-data)
+
+          ;; Apply owl-datalog reasoning
+          db-reasoned @(fluree/reason db-with-data :owl-datalog)]
+
+      (testing "Instance with all intersection components should be inferred as DrugProduct"
+        (is (= #{"ex:DrugProduct" "ex:ManufacturedItem" "ex:SubstanceDefinition"}
+               (set @(fluree/query db-reasoned
```
Comment (dpetran) [inline test/fluree/db/reasoner/owl_datalog_test.clj]: Do we need to coerce the results into a set? By my read the output should be deterministic.

File: test/fluree/db/reasoner/owl_datalog_test.clj [hunk 8]
```diff
@@ -0,0 +1,401 @@
...
+                         "insert"   [{"@id"        "ex:john"
+                                      "ex:hasChild" {"@id" "ex:mary"}}
+                                     {"@id"   "ex:mary"
+                                      "@type" "ex:Person"}]}
+
+          db-with-data @(fluree/update db-with-ontology instance-data)
+          db-reasoned @(fluree/reason db-with-data :owl-datalog)]
+
+      (testing "Individual with property pointing to correct type should be inferred as Parent"
+        (is (contains? (set @(fluree/query db-reasoned
```
Comment (dpetran) [inline test/fluree/db/reasoner/owl_datalog_test.clj]: Same thing here, there should only be a single result type, so no need to put it in a set.

**Review by dpetran — APPROVED** [review]
:owl:

File: docs/owl-datalog.md (hunk 1 above)
Comment (bplatz) [reply docs/owl-datalog.md]: Update doc to clarify.
